File: 2024 August/Edit Distance/main.cpp
// Link: https://www.geeksforgeeks.org/problems/edit-distance3702/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int editDistance(string str1, string str2)
    {
        vector<vector<int>> dp(str1.length() + 1, vector<int>(str2.length() + 1, 0));
        for (int j = 0; j <= str2.length(); j++)
        {
            dp[str1.length()][j] = str2.length() - j;
        }

        for (int i = 0; i <= str1.length(); i++)
        {
            dp[i][str2.length()] = str1.length() - i;
        }

        for (int i = str1.length() - 1; i >= 0; i--)
        {
            for (int j = str2.length() - 1; j >= 0; j--)
            {
                int ans = 0;
                if (str1[i] == str2[j])
                {
                    ans = dp[i + 1][j + 1];
                }
                else
                {
                    int insert = 1 + dp[i][j + 1];
                    int del = 1 + dp[i + 1][j];
                    int rep = 1 + dp[i + 1][j + 1];
                    ans = min(insert, min(del, rep));
                }
                dp[i][j] = ans;
            }
        }

        return dp[0][0];
    }
};

int main()
{
    Solution obj;
    string str1 = "geek";
    string srt2 = "gesek";
    cout << obj.editDistance(str1, srt2) << endl;

    str1 = "gfg";
    srt2 = "gfg";
    cout << obj.editDistance(str1, srt2) << endl;
    return 0;
}

File: 2024 August/Edit Distance/main.py
# Link: https://www.geeksforgeeks.org/problems/edit-distance3702/1


class Solution:
    def editDistance(self, str1, str3):
        dp = [[0] * (len(str3) + 1) for _ in range(len(str1) + 1)]

        for j in range(len(str3) + 1):
            dp[len(str1)][j] = len(str3) - j

        for i in range(len(str1) + 1):
            dp[i][len(str3)] = len(str1) - i

        for i in range(len(str1) - 1, -1, -1):
            for j in range(len(str3) - 1, -1, -1):
                if str1[i] == str3[j]:
                    dp[i][j] = dp[i + 1][j + 1]
                else:
                    insert = 1 + dp[i][j + 1]
                    delete = 1 + dp[i + 1][j]
                    replace = 1 + dp[i + 1][j + 1]
                    dp[i][j] = min(insert, delete, replace)

        return dp[0][0]


if __name__ == "__main__":
    obj = Solution()
    str1 = "geek"
    str2 = "gesek"
    print(obj.editDistance(str1, str2))

    str1 = "gfg"
    str2 = "gfg"
    print(obj.editDistance(str1, str2))
